data_structures/src/chain.rs

```diff
@@ -2902,6 +2900,111 @@ impl ReputationEngine {
     pub fn clear_threshold_cache(&self) {
         self.threshold_cache.borrow_mut().clear_threshold_cache()
     }
+
+    /// Get ARS keys ordered by reputation. If tie, order by pkh.
+    pub fn get_rep_ordered_ars_list(&self) -> Vec<PublicKeyHash> {
+        self.ars
+            .active_identities()
+            .cloned()
+            .sorted_by_key(|&pkh| std::cmp::Reverse((self.trs.get(&pkh), pkh)))
+            .collect()
+    }
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+fn magic_line(x: f64, m: f64, k: f64) -> u32 {
+    let res = m * x + k;
+    if res < 0_f64 {
+        0
+    } else {
+        res.round() as u32
+    }
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_precision_loss)]
+fn trapezoid_reputation(
+    sorted_rep: Vec<PublicKeyHash>,
+    trs: &TotalReputationSet<PublicKeyHash, Reputation, Alpha>,
+) -> (u64, Vec<u32>, HashMap<PublicKeyHash, u32>) {
+    let mut total_active_rep = 0;
+    let filtered_identities = sorted_rep
+        .iter()
+        .filter_map(|pkh| {
+            let rep = trs.get(pkh).0;
+            if rep != 0 {
+                total_active_rep += u64::from(rep);
+                Some(rep)
+            } else {
+                None
+            }
+        })
+        .count();
+
+    let mut hm = HashMap::default();
+
+    let trapezoid_rep = if filtered_identities != 0 {
+        let average = total_active_rep as f64 / filtered_identities as f64;
+        let minimum = f64::from(trs.get(&sorted_rep[filtered_identities - 1]).0);
+
+        // Parameters of the line y = mx + k
+        let k = average;
+        let m = (minimum - k) / ((filtered_identities as f64) - 1_f64);
+
+        let mut rep_triangle = 0;
+        let mut rep_triangle_vec = vec![];
+        for (i, _pkh) in sorted_rep.iter().enumerate() {
+            if i < filtered_identities {
+                let new_rep = magic_line(i as f64, m, k);
+                rep_triangle += new_rep;
+                rep_triangle_vec.push(new_rep);
+            } else {
+                rep_triangle_vec.push(0);
+            }
+        }
+
+        let diff_rep = total_active_rep as u32 - rep_triangle;
+        let offset = diff_rep / (filtered_identities as u32);
+        let diff_extra = (diff_rep as usize) % filtered_identities;
+
+        total_active_rep = 0;
+        rep_triangle_vec
+            .iter()
+            .enumerate()
+            .map(|(i, rep)| {
+                let new_rep = if i < diff_extra {
+                    rep + offset + 2
+                } else if i < filtered_identities {
+                    rep + offset + 1
+                } else {
+                    rep + 1
+                };
+
+                // Include the modified reputation in the hashmap
+                hm.insert(sorted_rep[i], new_rep);
+
+                new_rep
+            })
```

Comment:
- Please document this closure so that alternative implementations have a clear reference on what the algorithm is doing.
- This closure can be refactored into a proper function that we can unit test.
- It does not seem obvious to me what we are doing with `diff_extra` at this point, which I sort of understand that represents the reputation points that are still not covered by neither of the triangle and the rectangle.